### PR TITLE
Parallelize sending the client state API-1440

### PR DESCRIPTION
**Pull request by srknzl**
The original description of the issue is: 
```
Currently, we are sending the client state to the cluster one by one, for

- user code deployment

- compact schemas

- query caches

- proxies

It might be a good idea to investigate this, and try to come up with a solution that would handle all of this in one client message, as every request means extra latency during the client reconnects and cluster switches.
```

After some research we understood that it is not feasible to combine these into a single message. To see why, see the [Jira](https://hazelcast.atlassian.net/browse/API-1440) issue. Instead, I used client's task schedulor to run the invocations in parallel. Before my change, we were waiting each future and then sending another client message afterwards. Now, we will send all client messages in parallel. This should make the client state sending faster.

Breaking changes (list specific methods/types/messages):
* None

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by srknzl** [comment]
I have run a benchmark. You can find the code I used in these repos:  

https://github.com/srknzl/deleteme-benchmark-client
https://github.com/srknzl/deleteme-benchmark-server

so basically, I calculated the `Mode.SampleTime` to send the client state of 
10 user code + 100 compact schemas + 100 query cache + 10000 Imap proxies + 1000 ICaches
using jmh. The benchmark code has a setup code which includes creating the client and setting up the client state. The only thing I measured is the sendClientToCluster method. The benchmark has a teardown logic which clears all the state created, and shuts down the client.

I compared the code in master vs this PR's change.

**Environment**

2 machines used in aws eu-central-1 region. 1 for running a member, the other for running the benchmark. t2-medium machines are used.


**Benchmark results:**

```
Benchmark                                                Mode  Cnt  Score   Error  Units
MyBenchmark.testMethod                                 sample  312  0.831 ± 0.016   s/op
MyBenchmark.testMethod:testMethod·p0.00                sample       0.722           s/op
MyBenchmark.testMethod:testMethod·p0.50                sample       0.808           s/op
MyBenchmark.testMethod:testMethod·p0.90                sample       0.949           s/op
MyBenchmark.testMethod:testMethod·p0.95                sample       0.992           s/op
MyBenchmark.testMethod:testMethod·p0.99                sample       1.104           s/op
MyBenchmark.testMethod:testMethod·p0.999               sample       1.386           s/op
MyBenchmark.testMethod:testMethod·p0.9999              sample       1.386           s/op
MyBenchmark.testMethod:testMethod·p1.00                sample       1.386           s/op
MyBenchmark.testMethodMaster                           sample  438  0.588 ± 0.015   s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.00    sample       0.480           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.50    sample       0.561           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.90    sample       0.715           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.95    sample       0.799           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.99    sample       0.867           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.999   sample       1.376           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p0.9999  sample       1.376           s/op
MyBenchmark.testMethodMaster:testMethodMaster·p1.00    sample       1.376           s/op
```

File: hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -907,10 +911,30 @@ public void waitForInitialMembershipEvents() {
     }
 
     public void sendStateToCluster() throws ExecutionException, InterruptedException {
-        userCodeDeploymentService.deploy(this);
-        schemaService.sendAllSchemas();
-        queryCacheContext.recreateAllCaches();
-        proxyManager.createDistributedObjectsOnCluster();
+        CompletionService<Void> completionService = new ExecutorCompletionService<>(getTaskScheduler());
+        List<Callable<Void>> tasks = new ArrayList<>();
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/impl/clientside/HazelcastClientInstanceImpl.java:915]: Why do we need to store them in an array list first?

File: hazelcast/src/main/java/com/hazelcast/client/map/impl/querycache/ClientQueryCacheContext.java [hunk 2]
```diff
@@ -147,20 +150,29 @@ public Address getThisNodesAddress() {
         throw new UnsupportedOperationException();
     }
 
-    public void recreateAllCaches() {
+    public void recreateAllCaches() throws ExecutionException, InterruptedException {
+        CompletionService<Void> completionService = new ExecutorCompletionService<>(client.getTaskScheduler());
```
Comment (mdumandag) [inline hazelcast/src/main/java/com/hazelcast/client/map/impl/querycache/ClientQueryCacheContext.java:154]: If I were to choose, I would get rid of the parallelization in here and in ClientCacheProxyFactory. I feel like they are rarely used and the benefits will be small anyways. Also, it makes reasoning about the flow much easier

**Comment by srknzl** [comment]
@ihsandemir I am closing as won't do because the improvement is not valid for every use case as @mdumandag pointed out.

**Pull request closed without merging**